### PR TITLE
ACG provisioning platform API spec

### LINKED ISSUE #25899: [Azure Communications Gateway ] API Review
New API Review meeting has been requested.

**Service Name**: Azure Communications Gateway 
**Review Created By**: Richard Lupton
**Review Date**: 10/18/2023 09:00 AM PT
**Onboarding Record**: 
**PR**: 
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: This API is being added as a data plane API to Azure Communications Gateway (ACG) to allow telco operators to provision numbers to use Teams Operator Connect and DR (and later other providers like zoom). It will provide a unified API for provisioning these services.

 
The API at present is quite simple, it describes enterprise accounts (customers of the operator), who are to be provisioned for different voice services, and numbers under each account with the services that they use. This provides data which the communications gateway can use to route calls correctly.

The common case API requests are likely GETs and PUTs of numbers

GET / PUT api/v1.0/{account_name}/{number}
 

with operations on accounts / batch number upload being somewhat less frequent.

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=96e51599-8058-ee11-be6f-000d3a341a60)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by rlupton20-msft**
# Data Plane API - Pull Request

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue: https://github.com/Azure/azure-rest-api-specs/issues/25899

Is this review for (select one):

- [ ] a private preview
- [ ] a public preview
- [x] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document: can provide details on request.
* Previous API Spec Doc:
* Updated paths:

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

### CODE REVIEW

**Comment by weshaggard** [comment]
@lmazuel any idea who would be the best reviewer for this?

**Comment by rlupton20-msft** [comment]
[README.md](https://github.com/Azure/azure-rest-api-specs/files/12870186/README.md)

@tg-msft @mikekistler I've attached some docs detailing how we expected users to connect and use this API. This should cover the core concepts (accounts and numbers), and give a bunch of likely usage scenarios (with responses etc).

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/v1.0/swagger.json [hunk 1]
```diff
@@ -0,0 +1,1052 @@
...
+  "info": {
+    "title": "Microsoft.VoiceServices.ACG.ProvisioningPlatform.API",
+    "version": "v1.0"
+  },
+  "securityDefinitions": {
+    "OAuth2": {
+      "type": "oauth2",
+      "description": "Azure Active Directory OAuth2 Flow.",
+      "flow": "implicit",
+      "authorizationUrl": "https://login.microsoftonline.com/common/oauth2/authorize",
```
Comment (rlupton20-msft) [inline specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/v1.0/swagger.json:12]: This is the flow that we need: https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow#get-a-token but it's swagger isn't compatible with tenant specific token URLs.

Please advise how we should handle this.

**Review by mikekistler — CHANGES_REQUESTED** [review]
See [comments in review issue](https://github.com/Azure/azure-rest-api-specs/issues/25899#issuecomment-1768975450)

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/v1.0/swagger.json (hunk 1 above)
Comment (rlupton20-msft) [reply specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/v1.0/swagger.json]: Believe this is now resolved.

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json [hunk 2]
```diff
@@ -0,0 +1,1015 @@
...
+    {
+      "OAuth2": [
+        "https://func-voiceservice-rp-prod-eastuseuap.azurewebsites.net/.default"
+      ]
+    }
+  ],
+  "paths": {
+    "/account/{accountName}/numbers": {
+      "get": {
+        "operationId": "Account_GetNumbers",
```
Comment (mikekistler) [inline specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:28]: This operation and several others are missing the required "api-version" query parameter.
```
>spectral lint $f | grep api-version
  32:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./account/{accountName}/numbers.get.parameters
  98:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./account/{accountName}/numbers/batch.post.parameters
 154:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./account/{accountName}/numbers/{phoneNumber}.delete.parameters
 208:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./account/{accountName}/numbers/{phoneNumber}.get.parameters
 268:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./account/{accountName}/numbers/{phoneNumber}.put.parameters
 344:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./accounts.get.parameters
 397:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./accounts/{accountName}.delete.parameters
 441:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./accounts/{accountName}.get.parameters
 499:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./accounts/{accountName}.put.parameters
 565:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./numbers.get.parameters
```

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json [hunk 3]
```diff
@@ -0,0 +1,1015 @@
...
+          },
+          {
+            "format": "int32",
+            "in": "query",
+            "name": "skip",
+            "type": "integer",
+            "description": "An offset into the collection of the first record to be returned."
+          }
+        ],
+        "responses": {
```
Comment (mikekistler) [inline specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:58]: Operations with a pageable response should specify the `x-ms-pageable` extension.

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json [hunk 4]
```diff
@@ -0,0 +1,1015 @@
...
+          "200": {
+            "description": "Success",
+            "schema": {
+              "$ref": "#/definitions/NumberResponseDtoPaginatedListDto"
```
Comment (mikekistler) [inline specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:62]: Schema names should be clear and concise.
```suggestion
              "$ref": "#/definitions/NumberList"
```

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json [hunk 5]
```diff
@@ -0,0 +1,1015 @@
...
+            "type": "string",
+            "description": "Name of the account. Can only contain letters, numbers, underscores and dashes. Can be up to 100 characters in length."
+          },
+          {
+            "in": "body",
+            "name": "body",
+            "schema": {
+              "$ref": "#/definitions/BatchNumbersPutDto"
+            }
+          }
```
Comment (mikekistler) [inline specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:115]: Should this body parameter be marked as required?

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json [hunk 6]
```diff
@@ -0,0 +1,1015 @@
...
+            "$ref": "./examples/GetNumber.json"
+          }
+        },
+        "tags": [
+          "Number"
+        ],
+        "description": "Gets the details of a number assigned to an account, including the services for which the number is enabled."
+      },
+      "put": {
+        "operationId": "Number_Put",
```
Comment (mikekistler) [inline specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:261]: The operationId should be descriptive of the operation and not the underlying HTTP method used.
```suggestion
        "operationId": "Number_CreateOrReplace",
```

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json [hunk 7]
```diff
@@ -0,0 +1,1015 @@
...
+          {
+            "in": "header",
+            "name": "If-Match",
+            "type": "string"
+          },
+          {
+            "in": "header",
+            "name": "If-None-Match",
+            "type": "string"
+          },
```
Comment (mikekistler) [inline specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:298]: All parameters should have a description for the benefit of the generated documentation.

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json [hunk 8]
```diff
@@ -0,0 +1,1015 @@
...
+          {
+            "in": "body",
+            "name": "body",
+            "schema": {
+              "$ref": "#/definitions/NumberDetailsDto"
+            }
+          }
```
Comment (mikekistler) [inline specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:305]: Should this body parameter be marked required?

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json [hunk 9]
```diff
@@ -0,0 +1,1015 @@
...
+            "$ref": "./examples/GetAccount.json"
+          }
+        },
+        "tags": [
+          "Account"
+        ],
+        "description": "Gets the details of the specified account. An account represents a customer."
+      },
+      "put": {
+        "operationId": "Account_Put",
```
Comment (mikekistler) [inline specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:492]: The operationId should be descriptive of the operation and not the underlying HTTP method used.
```suggestion
        "operationId": "Account_CreateOrReplace",
```

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json [hunk 10]
```diff
@@ -0,0 +1,1015 @@
...
+      },
+      "type": "object",
+      "description": "Details of the account."
+    },
+    "AccountResponseDto": {
+      "additionalProperties": false,
+      "properties": {
+        "etag": {
+          "type": "string",
+          "x-nullable": true,
```
Comment (mikekistler) [inline specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:637]: Avoid the use of `x-nullable`.

**Review by mikekistler — CHANGES_REQUESTED** [review]
I flagged a few more issues to resolve.

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json (hunk 2 above)
Comment (rlupton20-msft) [reply specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:28]: Ah, apologies - I'd created the parameter description but not included it on the endpoints - fixed.

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json (hunk 10 above)
Comment (rlupton20-msft) [reply specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json]: Removed.

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json (hunk 6 above)
Comment (rlupton20-msft) [reply specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:276]: Done. Also checked the other operationsIds (apart from those flagged, the batch endpoint also needed some tweaking).

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json (hunk 9 above)
Comment (rlupton20-msft) [reply specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:522]: Updated. Checked other operationIds too.

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json (hunk 5 above)
Comment (rlupton20-msft) [reply specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json]: Yes - went over all of the requests with bodies and fixed this.

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json (hunk 7 above)
Comment (rlupton20-msft) [reply specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json]: Added documentation.

**Comment by rlupton20-msft** [comment]
Addressed the new issues, and took over the lint output and fixed up things in the same vein (and other things I thought worthwhile). Thanks for reviewing! Back over to you.

**Review by mikekistler — CHANGES_REQUESTED** [review]
Just a few more minor issues to resolve.

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json [hunk 11]
```diff
@@ -0,0 +1,1063 @@
...
+        "Provisioned",
+        "NotProvisioned",
+        "Inconsistent"
+      ],
+      "type": "string",
+      "description": "Status of DNS configuration. Azure Communications Gateway generates DNS records to Microsoft 365 to verify that the customer-specific domain name is associated with the correct customer tenant.",
+      "x-ms-enum": {
+        "name": "SubdomainProvisioningStatus",
+        "modelAsString": true
+      }
```
[the remark this answers is not in the document]
Comment (rlupton20-msft) [reply specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json]: Done.

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json [hunk 12]
```diff
@@ -0,0 +1,1063 @@
...
+          }
+        },
+        "innererror": {
+          "description": "An object containing more specific information about the error.",
+          "$ref": "#/definitions/InnerError"
+        },
+        "additionalProperties": {
+          "type": "object",
+          "description": "Additional properties that can be provided about the error."
+        }
```
[the remark this answers is not in the document]
Comment (rlupton20-msft) [reply specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json]: Done.

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json [hunk 13]
```diff
@@ -0,0 +1,1063 @@
...
+            "in": "body",
+            "name": "body",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/NumberDetails"
+            },
+            "description": "New details of the number to create or replace."
+          }
+        ],
+        "responses": {
```
[the remark this answers is not in the document]
Comment (rlupton20-msft) [reply specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:331]: Done.

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json [hunk 14]
```diff
@@ -0,0 +1,1063 @@
...
+      "required": [
+        "name"
+      ],
+      "type": "object",
+      "description": "Details of the account as returned on a GET request."
+    },
+    "AccountsList": {
+      "additionalProperties": false,
+      "properties": {
+        "records": {
```
[the remark this answers is not in the document]
Comment (rlupton20-msft) [reply specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json]: Changed to `value`.

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json [hunk 15]
```diff
@@ -0,0 +1,1093 @@
...
+  "info": {
+    "title": "Microsoft.VoiceServices.ACG.ProvisioningPlatform.API",
+    "version": "2023-10-01"
+  },
+  "securityDefinitions": {
+    "OAuth2": {
+      "type": "oauth2",
+      "description": "Azure Active Directory OAuth2 Flow.",
+      "flow": "application",
+      "tokenUrl": "https://login.microsoftonline.com/common/oauth2/v2.0/token",
```
Comment (rlupton20-msft) [inline specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:12]: I believe our service uses a tenant specific authentication endpoint for this flow. How do we express that here?

Comment (rlupton20-msft) [reply specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:12]: Our flow looks like this, https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-client-creds-grant-flow#get-a-token but as far as I can tell CI rejects any templating, and I've been unable to find any docs on how this is meant to be done.

**Review by mikekistler — APPROVED** [review]
Looks good! 👍 

Thanks for the fixes!

File: specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json (hunk 15 above)
Comment (mikekistler) [reply specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:12]: I think the SDK generators all have ways to provide more specific information on authentication, so what you have here should be fine.

Comment (rlupton20-msft) [reply specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:12]: Thanks - that makes sense.

**Review by JeffreyRichter — APPROVED** [review]

**Comment by mikekistler** [comment]
Not sure why this PR has no assignee -- it looks like none was ever assigned.

@tjprescott this PR is ready to merge.

**Pull request merged**